File: code/packages/elixir/repl/lib/coding_adventures/repl/language.ex
defmodule CodingAdventures.Repl.Language do
  @moduledoc """
  The Language behaviour — the heart of the REPL's pluggability.

  ## What Is a Language?

  In a REPL, a *language* is anything that can take a string of text and
  produce a result. That result is one of three things:

  1. `{:ok, String.t()}` — evaluation succeeded and there is a value to show.
  2. `{:ok, nil}` — evaluation succeeded but there is nothing to display
     (e.g. an assignment that produces no value, like `x = 5` in Python).
  3. `{:error, String.t()}` — evaluation failed with a message to display.
  4. `:quit` — the user wants to exit the session.

  ## Why a Behaviour?

  Using a behaviour (Elixir's equivalent of a protocol or interface) lets the
  REPL loop remain completely agnostic about the language it is evaluating.
  You can plug in:

  - A trivial echo language (for testing and demos)
  - A mathematical expression evaluator
  - A full Elixir interpreter
  - A Brainfuck engine
  - Anything else that can map `String.t() → result`

  The loop only knows about this contract. Nothing more.

  ## Implementing a Language

  ```elixir
  defmodule MyMathLanguage do
    @behaviour CodingAdventures.Repl.Language

    @impl true
    def eval(":quit"), do: :quit
    def eval(input) do
      case Integer.parse(input) do
        {n, ""} -> {:ok, Integer.to_string(n * 2)}
        _       -> {:error, "not a number: \#{input}"}
      end
    end
  end
  ```
  """

  # ---------------------------------------------------------------------------
  # Callback definition
  #
  # eval/1 receives the raw input string (already stripped of its trailing
  # newline by the loop). It must return one of the four tagged values above.
  # ---------------------------------------------------------------------------

  @doc """
  Evaluate a single line of input in this language.

  ## Parameters

  - `input` — the raw text the user typed, newline already stripped.

  ## Returns

  - `{:ok, value}` — success; `value` is a printable string or `nil`.
  - `{:error, message}` — failure; `message` describes what went wrong.
  - `:quit` — the user requested an end to the session.
  """
  @callback eval(input :: String.t()) ::
              {:ok, String.t() | nil}
              | {:error, String.t()}
              | :quit
end

File: code/packages/ruby/repl/lib/coding_adventures/repl/language.rb
# frozen_string_literal: true

# ============================================================================
# Language — the pluggable evaluator interface
# ============================================================================
#
# Every REPL needs a "language" — something that takes a string of user input
# and evaluates it, returning either a result, an error, or a signal to quit.
#
# This module documents the interface that any language implementation must
# satisfy. Ruby uses duck typing, so there is no enforcement: any object that
# responds to `eval` with the correct return contract can be used. The module
# is here for documentation and as an optional mixin.
#
# ## The contract
#
# `eval(input)` receives the raw string the user typed and must return one of:
#
#   [:ok, String]   — evaluation succeeded; the String is the output to print
#                     (may be nil if there is nothing to show, e.g. an assignment)
#   [:error, String] — evaluation failed; the String is the error message
#   :quit            — the user asked to exit; the loop should terminate cleanly
#
# ## Design note: why a tagged union return?
#
# Ruby doesn't have sum types, but we can simulate them with tagged arrays.
# Returning `[:ok, value]` vs `[:error, message]` vs `:quit` gives callers a
# clear, exhaustive set of cases to pattern-match (via `case/in`). This is far
# safer than raising exceptions across thread boundaries or relying on nil.
#
# Compare with Elixir's `{:ok, value} | {:error, reason}` — same idea, just
# in Ruby's syntax.

module CodingAdventures
  module Repl
    # Language is the evaluator interface.
    #
    # Any object that implements `eval(input)` according to the contract above
    # can serve as a language backend for the REPL loop.
    #
    # Implementors may include this module to signal intent, but it is not
    # required — duck typing suffices.
    module Language
      # Evaluate a single line of user input.
      #
      # @param input [String] the raw text entered by the user
      # @return [Array(:ok, String|nil)]   evaluation succeeded
      # @return [Array(:error, String)]    evaluation failed with a message
      # @return [:quit]                    the session should end
      def eval(input)
        raise NotImplementedError, "#{self.class}#eval must be implemented"
      end
    end
  end
end
